lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -34,21 +34,164 @@ use constant MAX_JOB_ALLOCATION => $ENV{OPENQA_SCHEDULER_MAX_JOB_ALLOCATION} //
+
+    # find jobs
+    my @jobs;
+    my $job_ids_str = join(', ', @$job_ids);
+    try {
+        @jobs = $schema->resultset('Jobs')->search({id => {-in => $job_ids}});
+    }
+    catch {
+        log_debug("Failed to retrieve jobs ($job_ids_str) in the DB, reason: $_");
+    };
+    my $actual_job_count = scalar @jobs;
+    if ($actual_job_count != scalar @$job_ids) {
+        log_debug("Failed to retrieve jobs ($job_ids_str) in the DB, reason: only got $actual_job_count jobs");
+        next;
+    }
+
+    # check whether the jobs are still scheduled
+    if (my @skipped_jobs = grep { $_->state ne SCHEDULED } @jobs) {
+        log_debug('Job ' . $_->id . ' no longer scheduled, skipping') for @skipped_jobs;
+        next;
+    }
+
+    # check whether the jobs have no worker assigned yet (so jobs already pulled as chained children are not
+    # scheduled twice)
+    if (my @skipped_jobs = grep { defined $_->assigned_worker_id } @jobs) {
+        log_debug('Job ' . $_->id . ' has already a worker assigned, skipping') for @skipped_jobs;
+        next;
+    }
+
+    # assign the jobs to the worker and then send the jobs to the worker
+    # note: The $worker->update(...) is also done when the worker sends a status update. That is
+    #       required to track the worker's current job when assigning multiple jobs to it. We still
+    #       need to set it here immediately to be sure the scheduler does not consider the worker
+    #       free anymore.
+    my $res;
+    try {
+        if ($actual_job_count > 1) {
+            my %worker_assignment = (
+                state              => ASSIGNED,
+                t_started          => undef,
+                assigned_worker_id => $worker_id,
+            );
+            $schema->txn_do(
+                sub {
+                    $_->update(\%worker_assignment) for @jobs;
+                    $worker->set_current_job($jobs[0]);
+                });
+            $res
+            = $self->_assign_multiple_jobs_to_worker(\@jobs, $worker, $directly_chained_job_sequence, $job_ids);
+        }
+        else {
+            $schema->txn_do(
+                sub {
+                    $jobs[0]->set_assigned_worker($worker);
+                    $worker->set_current_job($jobs[0]);
+                });
+            $res = $jobs[0]->ws_send($worker);
+        }
+        die "Failed contacting websocket server over HTTP" unless ref($res) eq "HASH" && exists $res->{state};
+    }
+    catch {
+        log_debug("Failed to send data to websocket, reason: $_");
+    };
+
+    if (ref($res) eq 'HASH' && $res->{state} && $res->{state}->{msg_sent} == 1) {
+        # note: That only means the websocket server could *start* sending the message but not that the message
+        #       has been received and acknowledged by the worker.
+        log_debug("Sent job(s) '$job_ids_str' to worker '$worker_id'");
+        push(@successfully_allocated, map { {job => $_, worker => $worker_id} } @$job_ids);
```

Comment:
this array is not even defined in this function. 

Together with Jane's question: how exactly did you test this?

Reply:
I have explained that already in https://github.com/os-autoinst/openQA/pull/3601#issuecomment-735447991